CALL_METHOD inline caches

kmod opened the pull request:
I was thinking about the second line of the following code snippet:
```
l = []
l.append(x)
```

Ideally we would emit code that looks something like
```
if (PyList_CheckExact(l))
  PyList_Append(l, x);
else
  slowpath();
```

However here's what we do currently:

```
LOAD_FAST:

LOAD_METHOD:
set lasti
test for tracing
jump if tracing
load l from frame
incref l
move l onto the value stack
adjust the value stack pointer
load the type
check the version tag
not-taken jump to guard failure
load the dict offset
test the dict offset
take jump to dict check success
load the constant attribute value into rax
incref it
put it onto the stack
copy l from rdi to rax

CALL_METHOD:
set lasti
test for tracing
jump if tracing
copy rax (l) onto the value stack
adjust the value stack
copy tstate to rdi [arg1]
copy value stack pointer to rsi [arg2]
set rdx [arg3] to zero (using an xor)
cmpq + sbb (adjust arg3 by one if a method is found)
call aot function

call_functionMethodDescr2
[this is abbreviated, it's quite a few assembly instructions]
push 5 regs
guard on oparg
fetch f->ob_type
cmp it to PyMethodDescr_Type
jne if failed
check tstate->tracing
load the vectorcall
check if its null (for tpcall style)
load method_vectorcall_noargs
check if its that
call _PyObject_RealIsSubclass
check for recursion limit being exceeded
load func->d_method->ml_meth
test if its null
call it
decref something
decref something else
return

adjust vsp back by 16
test if rax is null (exception)
je if exception (not taken)

POP_TOP
mov rax to rdi
decref rdi
je if deallocated
```

The basis of this PR is that we have an extremely good guess of the exact object that will be called in the CALL_METHOD bytecode, since it is stored in the inline cache for the corresponding LOAD_METHOD bytecode.

This PR specifically targets calling "method descriptors" (methods that are builtin fuctions) of builtin classes.

So with this PR the above simplifies to

```
LOAD_FAST

LOAD_METHOD
set lasti
test for tracing
jump if tracing
load l
incref l
store l to the value stack
adjust value stack pointer
check if this is list
jump for guard failure
mov list.append to rax
store rax to the value stack
mov l back to rax

LOAD_FAST

CALL_METHOD
set lasti
test for tracing
jump for tracing
load x
incref x
move x to the value stack
move rax (l) to the value stack
adjust the value stack
load tstate->use_tracing
check if tstate->use_tracing is nonzero
jump if it's nonzero
load the method object from the value stack
check if its null
jump if its null
load self (l) from the value stack
check if its type is list
jump if its not a list
load x from the value stack into rsi (arg2)
call list_append
load x from the value stack
decref it
jump if dealloc
load al from the value stack
decref it
jump if dealloc
adjust the value stack
check for an exception
jump for exception

POP_TOP
move the result from rax to rdi
decref it
jump if deallocd        
```

Which includes the following optimizations:
- no more incref+decref of list.append (it's immortal now)
- Instead of checking the version tag and then the dictoffset, we just check if the type is list
- inline call_functionMethodDescr2
- skip all the guards inside call_functionMethodDescr2 once we know the exact value of the callable

This results in a pretty large reduction in the overhead of these calls. Here are some timings of these kinds of functions -- this is the total time to call the function, including all the bookkeeping overhead and the actual work of the function itself:

```
list.clear: 6.8ns before, 3.6ns now
list.append: 12ns before, 10ns now
list.insert: 20ns before, 15ns now
dict.pop: 8ns before, 6ns now
str.startswith: 17ns before, 14ns now
```

Unfortunately the behavior of this PR seems pretty different on my two machines, which I presume is due to one being an AMD cpu and the other being an Intel CPU. This PR is not a ton of help on the AMD cpu, and sometimes seems to hurt it (seems to be due to extra icache pressure from the additional code). On my intel cpu it speeds up the macrobenchmarks by about 1%:

```
  system 31.98+-0.21              net +0.0%
b704af68 25.39+-0.03              net -20.6% Add some more trace examples
44c2a121 25.32+-0.04  -0.3%+-0.2% net -20.8% Add a new loadattr cache type
8b2b8964 25.30+-0.02  -0.1%+-0.1% net -20.9% Implement CALL_METHOD "hinting"
5e89aee4 25.17+-0.03  -0.5%+-0.1% net -21.3% Specialize for 1-arg calls as well
b38929d1 25.20+-0.02  +0.1%+-0.1% net -21.2% Add fastcall support to the method hints
709c8bac 25.21+-0.03  +0.0%+-0.1% net -21.2% Small makefile bug
c6ef5680 25.18+-0.03  -0.1%+-0.1% net -21.3% Cap how many decrefs we'll inline
8a644ce9 25.13+-0.02  -0.2%+-0.1% net -21.4% Create specialized versions of decref_array
6fe7c029 25.24+-0.02  +0.4%+-0.1% net -21.1% Test that exposes a segfault
0ab3d9bd 25.12+-0.04  -0.5%+-0.1% net -21.5% Add an additional guard to fix this bug
7711c177 25.11+-0.03  -0.1%+-0.1% net -21.5% Make builtin methods immortal to avoid inc/decrefs
c4b1b2a9 25.19+-0.02  +0.3%+-0.1% net -21.2% Handle VARARGS methods
daeb0440 25.10+-0.04  -0.4%+-0.1% net -21.5% Support method_call ics for methods that take keyw
2b07f84b 25.08+-0.05  -0.1%+-0.2% net -21.6% Now that we handle all method types, clean this up
51105afe 25.11+-0.01  +0.1%+-0.2% net -21.5% Add in a check for tracing
```

I checked one of the benchmarks and all the remaining calls to call_functionMethodDescr* are due to calling method descriptors of subclasses of builtin classes (such as OrderedDict), which I think is possible to handle but seems to be a pretty minor case.


Future work:
- We could do this same optimization with LOAD_GLOBAL + CALL_FUNCTION pairs
- I'm not sure how to do this, but if we were able to fuse the LOAD_METHOD and CALL_METHOD bytecodes together, then we could remove the duplicate guards, and get rid of a bunch of stack manipulations and refcounting operations. Unfortunately those two bytecodes are not next to each other except for 0-arg calls.
- There's still a lot of value-stack manipulation in the inline cache which I think could be deferred to the slowpath
- We know that list_append always returns None (or NULL), so we could skip decreffing it.

kmod committed:
Add a new loadattr cache type

Which is specialized for handling objects of builtin types.

This lets us know that many of the fields are constant, and lets
us generate a more compact inline cache.

Here's what it was for "[].clear" before this commit:

```
      18 LOAD_METHOD
movl   $0x12,0x68(%r13)
cmpq   $0x0,(%rbx)
jne    a6b
mov    0x168(%r13),%rdi
incq   (%rdi)
mov    %rdi,(%r12)
add    $0x8,%r12
mov    0x8(%rdi),%rsi
cmpq   $0x13,0x180(%rsi)
jne    a8b

mov    0x120(%rsi),%rsi
test   %rsi,%rsi
je     209
js     a8b
mov    (%rdi,%rsi,1),%rsi
test   %rsi,%rsi
je     209
cmpq   $0x0,0x18(%rsi)
jne    a8b

movabs $0x7f44f4ce2770,%rax
incq   (%rax)
mov    %rax,-0x8(%r12)
mov    %rdi,%rax
```

The middle section is for testing to make sure there's no dict on the
instance so that we don't have to worry about an instance attribute
shadowing the type attribute, and this section is removed by this commit.

todo: we could save the tp_dictoffset in the normal caches so that we don't have to
load it again at jit time.

In terms of readability I think this would have made more sense as a new variable
or two in the existing loadattr caches. But to keep the cache size down I created
a new type and reused some of the fields.

I suppose this will slightly slow down the non-jitted case but hopefully that's minimal.

kmod committed:
Implement CALL_METHOD "hinting"

We will often have good runtime information available about
the callable we will call with CALL_METHOD, since this value
will often by cached for the corresponding LOAD_METHOD call.

This information is more specific than what our aot traces guards on,
so if we can remember it we can create an inline cache that is
quite a bit better than the aot trace. (For example we can statically
resolve getting the vectorcall function, type checking, and arg number checking.)

I was originally hoping to fuse the LOAD_METHOD + CALL_METHOD together,
which makes sense to me because then we can directly flow the info
within the same guard section.
But since there can be many bytecodes in between the LOAD_METHOD and CALL_METHOD,
just store some hint information along with the guard that's required to
validate the hint.

todo:
- LOAD_GLOBAL + CALL_FUNCTION can probably also benefit from this.
- It feels like a more general system could probably pick up a lot
  of similar cross-bytecode opportunities like this.
- There is still a decent amount of overhead left that we could get rid
  of. For example shuffling things to-and-from the value stack,
  incref+decreffing pairs that are unnecessary, etc.
  - I think this might already be possible with smarter use of the deferred_vs functions

kmod committed:
Specialize for 1-arg calls as well

but oddly this is actually slower??

kmod committed:
Add fastcall support to the method hints

kmod committed:
Small makefile bug

kmod committed:
Cap how many decrefs we'll inline

I'm starting to doubt the reliability of the microbenchmark I'm using,
so the initial results that inlining the decrefs was very bad may be spurious.

But I think it's still a good idea to not blow up the code size too much.
We'll have to test this on macrobenchmarks.

kmod committed:
Create specialized versions of decref_array

kmod committed:
Add an additional guard to fix this bug

kmod committed:
Test that exposes a segfault

kmod committed:
Make builtin methods immortal to avoid inc/decrefs on them

more things could be made immortal, such as other attributes
of builtin types, as well as builtin functions (not methods)

kmod committed:
Handle VARARGS methods

kmod committed:
Support method_call ics for methods that take keywords

Since we're only doing ICs for callsites without keyword
arguments, it's easy enough to support calling methods
that take keywords: we just pass NULL for the list of
keywords.

kmod committed:
Now that we handle all method types, clean this up a bit

Can do the guarding in the common section

kmod committed:
Add in a check for tracing

kmod committed:
Add some more trace examples

for builtin method calls with more arguments

This should hopefully be somewhat moot with the upcoming
inline caching of these paths, but I noticed that we were
missing these traces so let's add them anyway.

kmod committed:
note

undingen commented on Python/aot_ceval_jit.c:
```diff
@@ -2246,6 +2309,133 @@ void* jit_func(PyCodeObject* co, PyThreadState* tstate) {
+                            | mov arg1, [vsp - 8 * num_vs_args] // callable
+                            | test arg1, arg1
+                            | je >1
+
```
Replacing this mov+test compinations with the following code should save two instructions. I guess this will likely not change perf but I think it makes the code also a bit more readable because it's clear that it's just a guard and later code is not reading the register.
```
                            | cmp dword [tstate + offsetof(PyThreadState, use_tracing)], 0
                            | jne >1

                            | cmp dword [vsp - 8 * num_vs_args], 0 // callable
                            | je >1
```

undingen replied:
good think you also catched my `cmp dword error` which should have been a `qword` like you wrote.

undingen commented on Python/aot_ceval_jit.c:
```diff
@@ -2246,6 +2309,133 @@ void* jit_func(PyCodeObject* co, PyThreadState* tstate) {
+                        void* funcptr = method->d_method->ml_meth;
+
+                        if (funcptr && _PyObject_RealIsSubclass((PyObject*)hint->type, (PyObject *)PyDescr_TYPE(method))) {
+                            wrote_inline_cache = 1;
```
if you already set `wrote_inline_cache=1` here the later assignments can be removed.

undingen commented on Python/aot_ceval_jit.c:
```diff
@@ -3129,6 +3332,14 @@ void* jit_func(PyCodeObject* co, PyThreadState* tstate) {
+    CallMethodHint *hint = Dst->call_method_hints;
+    Dst->call_method_hints = NULL;
+    while (hint) {
+        CallMethodHint *new_hint = hint->next;
```
I think this code to free the memory never gets reached because AFAIK all the CALL_METHODs remove an element from the list. If this is here to catch unpaired LOAD_METHOD/CALL_METHOD we require a `free(hint)` at the end of the IC.

undingen reviewed (commented):
Wow this is a very cool PR.
As said on the call seeing the overhead written down like you did make it look quite bad :-P glad you managed to remove some of it.
I have three small comments:
- some wrote_inline_cache=1 can be removed
- generate two instructions less per method ic
- missing free(hint)

I could also confirm about 1% speedup on make measure on my intel alder lake

kmod committed:
Use a single instruction instead of two here

kmod committed:
Clean up wrote_inline_cache=1 handling

I added this variable when it wasn't guaranteed we'd be able to write
out an inline cache, but now that we handle all cases we don't need to
track this within this case (method_descr type where the type is correct)
anymore.

kmod committed:
Properly free the hints

I added code to free them all at the end, but forgot that I
pop them as they are used. So free them after they are popped.

Also add some defensive coding in case we are run on improper bytecode
where the LOAD_METHOD and CALL_METHOD are not matched.

kmod commented:
Ah good catches! I pushed a couple new commits that address them.
